src/stochastic_search_iterator.jl

```diff
@@ -0,0 +1,167 @@
+-   `initial_temperature::Int64` = 1 
+An iterator over all possible expressions of a grammar up to max_depth with start symbol sym.
+"""
+Base.@kwdef mutable struct StochasticSearchEnumerator <: ExpressionIterator
+    grammar::ContextFreeGrammar
+    max_depth::Int64 = 5  # maximum depth of the program that is generated
+    examples::Vector{Example}
+    neighbourhood::Function
+    propose::Function
+    accept::Function
+    temperature::Function
+    cost_function::Function
+    start_symbol::Symbol
+    initial_temperature::Int64 = 1
+end
+
+Base.@kwdef struct IteratorState
+    current_program::RuleNode
+    current_temperature::Float32
+    best_program::RuleNode
+    best_program_cost::Float32
+end
+
+Base.IteratorSize(::StochasticSearchEnumerator) = Base.SizeUnknown()
+Base.eltype(::StochasticSearchEnumerator) = RuleNode
+
+function Base.iterate(iter::StochasticSearchEnumerator)
+    grammar, max_depth = iter.grammar, iter.max_depth
+    # sample a random node using start symbol and grammar
+    sampled_program = rand(RuleNode, grammar, iter.start_symbol, max_depth)
+    current_cost = calculate_cost(sampled_program, iter.cost_function, iter.examples, iter.grammar)
+    return (sampled_program, IteratorState(
+        current_program=sampled_program,
+        current_temperature=iter.initial_temperature,
+        best_program=sampled_program,
+        best_program_cost=current_cost))
+end
+
+
+"""
+The algorithm that constructs the iterator of StochasticSearchEnumerator. It has the following structure:
+
+1. get a random node location -> location,dict = neighbourhood(current_program)
+2. call propose on the current program getting a list of possbile replacements in the node location 
+3. iterate through all the possible replacements and perform the replacement in the current program 
+    4.  accept the new program by modifying the next_program or reject the new program
+5. return the new next_program
+"""
+function Base.iterate(iter::StochasticSearchEnumerator, current_state::IteratorState)
+    grammar, examples = iter.grammar, iter.examples
+    current_program = current_state.current_program
+    
+    current_cost = calculate_cost(current_program, iter.cost_function, examples, grammar)
+
+    new_temperature = iter.temperature(current_state.current_temperature)
+
+    # get the neighbour node location 
+    neighbourhood_node_location, dict = iter.neighbourhood(current_state.current_program, grammar)
+
+    # get the subprogram pointed by node-location
+    subprogram = get(current_program, neighbourhood_node_location)
+
+
+    @info "Start: $(rulenode2expr(current_program, grammar)), subexpr: $(rulenode2expr(subprogram, grammar)), cost: $current_cost
+            temp $new_temperature"
+
+    # propose new programs to consider. They are programs to put in the place of the node
+    possible_replacements = iter.propose(current_program, neighbourhood_node_location, grammar, iter.max_depth, dict)
```

Comment:
what if `propose` returns an iterator instead of a list? that would make it much more memory efficient.